### PR TITLE
Relocate python to python/prairielearn

### PR DESCRIPTION
**Pull request by reteps**
- Allows canonical imports from prairielearn

This is important if we ever want to:

- Have a pip installable prairielearn package
- Have more than one namespace for imports
- Use mkdocstrings for documenting this code

Changes I made:

1. move python to python/prairielearn
2. fixup all imports to start with `prairielearn.`

For example `from python_helper_sympy` ➡️ `from prairielearn.python_helper_sympy`

3. moved the internal code to `python/prairielearn/internal` to indicate that it may change in the future

4. exported `prairielearn.prairielearn` helpers to the global package (so that `import prairielearn as pl` still works for most cases)

5. Add shims for modules still used by courses

New folder structure

```
elements
python
- prairielearn
- tests
- zygote.py
```

### CODE REVIEW

File: apps/prairielearn/python/prairielearn/check_data_test.py
Comment (eliotwrobson) [inline apps/prairielearn/python/prairielearn/check_data_test.py:1]: Along with this change, I think it would make sense to move the test files in the Python folder to their own subdirectory (the directory is getting kinda cluttered)

Comment (reteps) [reply apps/prairielearn/python/prairielearn/check_data_test.py:1]: I think I will do that in a separate PR. I would also like to clearly delineate library code that instructors can use from internal code that may have breakages.

Comment (reteps) [reply apps/prairielearn/python/prairielearn/check_data_test.py:1]: Actually I did this in [8f66a5c](https://github.com/PrairieLearn/PrairieLearn/pull/11051/commits/8f66a5c29ac008fb2eecffb830a507fa6658879b)

**Comment by reteps** [comment]
Here are some numbers:

- check_data: 0 dependencies
- question_phases: 0 dependencies
- zygote: 0 deps
- zygote_utils: 0 deps
- traverse: 3 dependencies

- colors: many dependencies
- python_helper_sympy: many dependencies
- to_precision: many dependencies

**Comment by nwalters512** [comment]
> For example from `python_helper_sympy` ➡️ from `prairielearn.python_helper_sympy`

What's the backwards-compat story for course code that's already using plain `import python_helper_sympy`? If we're going to be making that available under a new import path, why not take the opportunity to improve the name? I never did like `python_helper_sympy` 🙃

**Comment by reteps** [comment]
> > For example from `python_helper_sympy` ➡️ from `prairielearn.python_helper_sympy`
> 
> What's the backwards-compat story for course code that's already using plain `import python_helper_sympy`? If we're going to be making that available under a new import path, why not take the opportunity to improve the name? I never did like `python_helper_sympy` 🙃

Agreed -- the name could be improved. If we want a shim:

```
zygote.py
prairielearn
     python_helper_sympy.py
python_helper_sympy/__init__.py - points to prairielearn.python_helper_sympy
```
and use some path hacks

**Comment by reteps** [comment]
I have now added said shims. So the only breakage is the 1 course using `traverse.py`.

**Comment by SethPoulsen** [comment]
> > For example from `python_helper_sympy` ➡️ from `prairielearn.python_helper_sympy`
> 
> What's the backwards-compat story for course code that's already using plain `import python_helper_sympy`? If we're going to be making that available under a new import path, why not take the opportunity to improve the name? I never did like `python_helper_sympy` 🙃

Were you hoping for something more descriptive? Like `prairielearn_sympy_wrapper` or something?

**Comment by reteps** [comment]
> > > For example from `python_helper_sympy` ➡️ from `prairielearn.python_helper_sympy`
> > 
> > 
> > What's the backwards-compat story for course code that's already using plain `import python_helper_sympy`? If we're going to be making that available under a new import path, why not take the opportunity to improve the name? I never did like `python_helper_sympy` 🙃
> 
> Were you hoping for something more descriptive? Like `prairielearn_sympy_wrapper` or something?

I propose dropping `prairielearn and python` from the name, something like `prairielearn.sympy_utils` or `prairielearn.sympy_helper` seems to fill well.

**Comment by nwalters512** [comment]
I was going to suggest `sympy_utils`, yeah. It's obviously Python (so that doesn't need to be in the name) and I like "utils" better than "helper".

**Review by SethPoulsen — COMMENTED** [review]
I'm not sure it makes sense to put elements under `python/`. They aren't just python, they are python, javascript, css, html.

**Comment by reteps** [comment]
> I'm not sure it makes sense to put elements under python/. They aren't just python, they are python, javascript, css, html.

I agree, and I have reverted back.

File: apps/prairielearn/python/test/check_data_test.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import pytest
-from check_data import check_data
+from prairielearn.internal.check_data import check_data
```
Comment (reteps) [inline apps/prairielearn/python/test/check_data_test.py:2]: I would like to either use `internal.check_data` or `_check_data`, if someone has a preference for the latter, I can change it.

**Comment by reteps** [comment]
@nwalters512 -- I want to redo the mkdocstrings PR, so I would love a round of feedback on this if you have time.

File: apps/prairielearn/python/test/check_data_test.py
Comment (nwalters512) [inline apps/prairielearn/python/test/check_data_test.py:1]: Why the switch to a `test` directory? At least in JS, we really like having `foo.ts` and `foo.test.ts` adjacent to each other, and I think that's true for Python as well.

File: apps/prairielearn/python/test/traverse_test.py [hunk 2]
```diff
@@ -1,4 +1,8 @@
-from traverse import ElementReplacement, traverse_and_execute, traverse_and_replace
+from prairielearn.internal.traverse import (
+    ElementReplacement,
+    traverse_and_execute,
+    traverse_and_replace,
+)
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/python/test/traverse_test.py:5]: Looking at this again, I guess this is just necessary to split an import across multiple lines.

File: apps/prairielearn/python/colors/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# This is a shim to prairielearn.colors and will be removed in the future.
+import sys
+
+sys.path.append("..")
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/colors/__init__.py:4]: This is not possible in python unless you are inside a module. This is definitely a "do not do" escape hatch, but there is no other way to maintain backwards compatibility here.

File: apps/prairielearn/python/prairielearn/core.py [hunk 4]
```diff
@@ -21,15 +21,16 @@
...
 import sympy
-import to_precision
-from colors import PLColor
 from numpy.typing import ArrayLike
 from pint import UnitRegistry
 from text_unidecode import unidecode
 from typing_extensions import NotRequired, assert_never
 
+import prairielearn.sympy_utils as phs
+import prairielearn.to_precision as to_precision
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/prairielearn/core.py:31]: We are importing the entire module, not a to_precision helper in the module.

File: apps/prairielearn/python/prairielearn/internal/question_phases.py [hunk 5]
```diff
@@ -7,12 +7,13 @@
...
 
 import lxml.html
-from check_data import Phase, check_data
-from traverse import traverse_and_execute, traverse_and_replace
 from typing_extensions import assert_never
 
-PYTHON_PATH = pathlib.Path(__file__).parent.parent.resolve()
-CORE_ELEMENTS_PATH = (PYTHON_PATH / "elements").resolve()
+from prairielearn.internal.check_data import Phase, check_data
+from prairielearn.internal.traverse import traverse_and_execute, traverse_and_replace
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/prairielearn/internal/question_phases.py:13]: Yes -- there are even ruff rules that check for this.

File: apps/prairielearn/python/prairielearn/sympy_utils.py [hunk 6]
```diff
@@ -389,7 +389,10 @@ def evaluate_with_source(
 ) -> tuple[sympy.Expr, str]:
     # Replace '^' with '**' wherever it appears. In MATLAB, either can be used
     # for exponentiation. In Python, only the latter can be used.
-    expr = pl.full_unidecode(greek_unicode_transform(expr)).replace("^", "**")
+    # TODO: Split up full_unicode in prairielearn.py to prevent a circular import
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/prairielearn/sympy_utils.py:392]: Mmm, I wasn't planning on it, but I could.

File: apps/prairielearn/python/prairielearn/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/python/prairielearn/core.py:31]: Let me rephrase: every single call in this file is `to_precision.to_precision(...)`. Why not import the function directly and then change the calls to just be `to_precision(...)`?

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/python/prairielearn/sympy_utils.py:392]: I won't block on this, but it'd be really nice to resolve this. If you don't, please open an issue to track this.

File: apps/prairielearn/python/test/check_data_test.py
Comment (nwalters512) [reply apps/prairielearn/python/test/check_data_test.py:1]: If/when we ship this to PyPi, can't we just use `MANIFEST.in` to exclude `test_*.py` files?

File: apps/prairielearn/python/prairielearn/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/prairielearn/core.py:31]: Oh -- good point!

File: apps/prairielearn/src/lib/assets.ts [hunk 7]
```diff
@@ -21,7 +21,9 @@ let publicHash: HashElementNode | null = null;
 const cachedPackageVersionHashes: Record<string, string> = {};
 
 async function computeElementsHash() {
-  elementsHash = await hashElement(path.join(APP_ROOT_PATH, 'elements'), { encoding: 'hex' });
+  elementsHash = await hashElement(path.join(APP_ROOT_PATH, 'elements'), {
+    encoding: 'hex',
+  });
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/src/lib/assets.ts]: Reverted in 38ba29c84353cd303e49826f6ef039f0554cf8e8.

File: apps/prairielearn/python/test/check_data_test.py
Comment (reteps) [reply apps/prairielearn/python/test/check_data_test.py:1]: Hmm, I'm still onboard with splitting up into a test directory. Don't we have `/src/tests/*.test.ts` anyways?

File: apps/prairielearn/python/prairielearn/internal/question_phases.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/python/prairielearn/internal/question_phases.py:13]: For posterity, it looks like that rule is https://docs.astral.sh/ruff/rules/relative-imports/. By default, it allows relative imports from siblings per PEP 8:

> However, explicit relative imports are an acceptable alternative to absolute imports, especially when dealing with complex package layouts where using absolute imports would be unnecessarily verbose:

File: apps/prairielearn/python/prairielearn/core.py [hunk 8]
```diff
@@ -21,15 +21,16 @@
...
-import python_helper_sympy as phs
 import sympy
-import to_precision
-from colors import PLColor
 from numpy.typing import ArrayLike
 from pint import UnitRegistry
 from text_unidecode import unidecode
 from typing_extensions import NotRequired, assert_never
 
+import prairielearn.sympy_utils as phs
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/prairielearn/core.py]: Changed in 8266ba4cfcff806807435eb6cbd1bae36d788380.

File: .gitignore [hunk 9]
```diff
@@ -110,3 +110,6 @@ packages/*/coverage
 
 ## d2 cache
 .cache
+
+## v8 crash reports
+isolate-*-v8.log
```
[the remark this answers is not in the document]
Comment (reteps) [reply .gitignore:115]: I was getting a couple crashes on my computer, especially since I have low ram usage.

File: apps/prairielearn/elements/pl-big-o-input/__init__.py
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/elements/pl-big-o-input/__init__.py:1]: A python module/package is created when it has a `__init__.py` file. This allows other pieces of code to import it. For the #11048 extension to work, we need all code to be inside a package. So these new files allow us to autodocument our elements in the documentation. They would also allow in the future better code-sharing between elements.

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/prairielearn/sympy_utils.py:392]: I'm not sure the best way to do this -- typically you would split up the full_unicode into a seperate file and use it in both core/sympy_utils. Are you fine starting a new file, `utils.py`?

File: apps/prairielearn/python/colors/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/python/colors/__init__.py:4]: Here's another question (a lot of this is me learning things about Python for the first time, thanks for bearing with me): what's the difference between `foo.py` and `foo/__init__.py`? That is, why couldn't we have some file `apps/prairielearn/python/colors.py` that's just the following:

```py
from prairielearn.colors import *
```

File: apps/prairielearn/elements/pl-big-o-input/__init__.py
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/elements/pl-big-o-input/__init__.py:1]: Can you save the addition of `__init__.py` until it's actually needed? AFAICT that's unrelated to this PR, unless I'm missing something.

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/python/prairielearn/sympy_utils.py:392]: Maybe `unicode_utils.py` to mirror `sympy_utils.py`?

File: apps/prairielearn/python/test/check_data_test.py
Comment (nwalters512) [reply apps/prairielearn/python/test/check_data_test.py:1]: Those are integration tests that test the server as a whole; they aren't testing any particular file. For unit tests that test specific functions in a given file, we conventionally always place those adjacent to the specific file under test. Our rationale is mostly that this makes tests easy to discover and encourages a culture of testing code since the tests are so visible.

I won't be dogmatic about this, but I'd like to hear more about why this is discouraged beyond just "you'll ship tests" (which AFAICT can be avoided both here and in JS, e.g. via `package.json#files` or `.npmignore`). I couldn't glean much from the SO post you linked, I just see an assertion that "[t]he very common directory structure for even a simple Python module seems to be to separate the unit tests into their own test directory".

File: apps/prairielearn/python/colors/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/colors/__init__.py:4]: That would be allowed. I suppose as a shim, if you like doing that better, I can change it. The reason I did an explicit directory was just to be consistent with the module idea.

File: apps/prairielearn/python/test/check_data_test.py
Comment (reteps) [reply apps/prairielearn/python/test/check_data_test.py:1]: Here is some considerations from pytest: https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html.

> For unit tests that test specific functions in a given file, we conventionally always place those adjacent to the specific file under test.

Hmm, I buy keeping it in the same directory with this logic.

> Our rationale is mostly that this makes tests easy to discover and encourages a culture of testing code

If we use this, they definitely should be in the same directory. I sometimes think lots of tests mixed in-between source make it harder to find what I am looking for. That is the main reason I would split it up. I don't view a lot of libraries that keep it together either, I'm not sure why other than as I mentioned earlier, ease of packaging.

File: apps/prairielearn/python/prairielearn/sympy_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/prairielearn/sympy_utils.py]: Resolved in f5375f338.

File: apps/prairielearn/python/colors/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/colors/__init__.py:4]: However, in a future PR, if I make the elements modules with the __init__.py files, I will need to put the colors.py in a module so it can be imported. Otherwise we will get "can't import outside module" errors.

File: apps/prairielearn/elements/pl-big-o-input/__init__.py
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/elements/pl-big-o-input/__init__.py:1]: Not really missing anything; I just thought it fit better here given that this PR is turning all our python code into proper "modules". I can revert these and put them in the #11048 replacement however.

[the remark this answers is not in the document]
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-big-o-input/__init__.py:1]: I'm with Nate, I think this should be reverted for now. IIRC, the element controller files can't even be imported cleanly until they get renamed because of the hyphen: https://github.com/PrairieLearn/PrairieLearn/issues/8823

[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/elements/pl-big-o-input/__init__.py:1]: Works for me! Reverted in 6df545ab6.

File: apps/prairielearn/python/test/check_data_test.py
Comment (eliotwrobson) [reply apps/prairielearn/python/test/check_data_test.py:1]: Most of the Python packages I see (and the ones I've personally worked on) have tests in a separate folder. I'm in favor of keeping the tests for all of the core files in a separate folder, but the tests for the elements with those individual elements (since they each have just one test file, and the element controller script + test file is a pair that doesn't need to interact with anything else).

File: .gitignore (hunk 9 above)
[the remark this answers is not in the document]
Comment (eliotwrobson) [reply .gitignore:115]: Not sure why but whenever I run this platform locally, my ram usage goes insane.

File: apps/prairielearn/python/colors/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/python/colors/__init__.py:4]: > Otherwise we will get "can't import outside module" errors.

I wasn't able to reproduce this:

- I added an `__init__.py` file to `elements/pl-number-input`
- I made the above change to use `python/colors.py` instead of `python/colors/__init__.py`
- I loaded `template/number-input/random` without errors

Can you tell me how you saw that error so I can reproduce for myself?

[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/colors/__init__.py:4]: My apologies; I just tried to recreate the error and I could not. I logged the `sys.path` during execution and didn't run into any issues, and `typecheck-python` / `test-python` came back clean.

sys.path:
```
['/Users/retep/repos/PrairieLearn/apps/prairielearn/elements/pl-external-grader-results', '/Users/retep/repos/PrairieLearn/apps/prairielearn/python', '/Users/retep/repos/PrairieLearn/apps/prairielearn/python', '/Users/retep/.local/share/uv/python/cpython-3.10.16-macos-aarch64-none/lib/python310.zip', '/Users/retep/.local/share/uv/python/cpython-3.10.16-macos-aarch64-none/lib/python3.10', '/Users/retep/.local/share/uv/python/cpython-3.10.16-macos-aarch64-none/lib/python3.10/lib-dynload', '/Users/retep/repos/PrairieLearn/.venv/lib/python3.10/site-packages']
```

^ This clearly indicates that it would be able to import it, no issues.

At some point in time I was getting those errors; though I don't remember when.

[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/colors/__init__.py]: In [a80e198](https://github.com/PrairieLearn/PrairieLearn/pull/11051/commits/a80e1983bc88a42bf4391f2da8207fba2f574794) I made the shims a single file. In [53ae351](https://github.com/PrairieLearn/PrairieLearn/pull/11051/commits/53ae3519708422d4083de16375bb914519c8cf38) I clarify how the element execution happens, hopefully clarifying how the module resolution would work in the future.

File: apps/prairielearn/python/test/check_data_test.py
Comment (reteps) [reply apps/prairielearn/python/test/check_data_test.py:1]: +1 on @eliotwrobson suggestion.

File: apps/prairielearn/python/colors.py [hunk 10]
```diff
@@ -1,153 +1,4 @@
-"""
```
Comment (reteps) [inline apps/prairielearn/python/colors.py:1]: We lose file history on these three files currently. Should I do a multi-stage PR so that we don't lose history @nwalters512 ?

Comment (nwalters512) [reply apps/prairielearn/python/colors.py:1]: If it's possible, I think that's very much preferable, there's a lot of code/history here, particularly in the Sympy stuff. How would you approach that?

Comment (reteps) [reply apps/prairielearn/python/colors.py:1]: I can revert back to the `dir/__init__.py` file (which has a different file path) so that this doesn't happen.

Comment (nwalters512) [reply apps/prairielearn/python/colors.py:1]: Sure, let's give that a shot. Then we can immediately follow with another PR to move things around (without losing history), right?

Comment (reteps) [reply apps/prairielearn/python/colors.py]: right.

Comment (reteps) [reply apps/prairielearn/python/colors.py]: Changed in [c28afcc](https://github.com/PrairieLearn/PrairieLearn/pull/11051/commits/c28afcc076ce2dfa8f376db0db8f42fc86e65bfe).

File: apps/prairielearn/python/colors/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/python/colors/__init__.py:5]: In the most recent commits, we went back to using `colors/__init__.py` here. Locally, I removed the `sys.path` mucking around and was still able to `import colors` in a question. Can you tell me how to reproduce whatever error you saw that necessitated modifying `sys.path`?

File: apps/prairielearn/python/test/check_data_test.py
Comment (nwalters512) [reply apps/prairielearn/python/test/check_data_test.py:1]: I'm ok with this. Resolving.

File: apps/prairielearn/python/colors.py (hunk 10 above)
Comment (nwalters512) [reply apps/prairielearn/python/colors.py]: We can continue discussion here: https://github.com/PrairieLearn/PrairieLearn/pull/11051#discussion_r1919025388

File: apps/prairielearn/python/colors/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/colors/__init__.py:5]: You are correct; this must have been leftover from before I had modified this field in `pyproject.toml`:

```
extraPaths = ["./apps/prairielearn/python"]
```

Since all imports are resolved from this directory because of the zygote execution, it won't fail. But I generally never like making an import that is situational: if you ever did that import, and your execution didn't start from `./apps/prairielearn/python`, it would fail.

[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/python/colors/__init__.py]: Anyways, this should be fixed almost immediately by a followup PR, so I just removed it in e9205f26b870c6feb845d76cfecba340694f15ae.

File: apps/prairielearn/python/zygote.py [hunk 11]
```diff
@@ -68,6 +68,7 @@
 
 # Pre-load commonly used modules
 sys.path.insert(0, os.path.abspath("../question-servers/freeformPythonLib"))
```
Comment (nwalters512) [inline apps/prairielearn/python/zygote.py:70]: AFAICT this path simply doesn't exist anymore. Can we remove this?

File: apps/prairielearn/python/README.md [hunk 12]
```diff
@@ -0,0 +1,5 @@
+# Python Directory
+
+This directory contains all the shared python code. Elements can import from this directory to use shared code. You should not import from `test` or from `prairielearn.internal`, as they are not guaranteed to have a stable API.
+
+All element code is run in the context of zygote.py, so it will be run from this directory.
```
Comment (nwalters512) [inline apps/prairielearn/python/README.md:5]:
```suggestion
All element code is run in the context of `zygote.py`, so it will be run from this directory.
```

The fact that questions are executed by the Zygote isn't directly connected to why things can e.g. import `prairielearn`. Questions are executed with their working directory as the question. The zygote just adds its own working directory to Python's path:

https://github.com/PrairieLearn/PrairieLearn/blob/b0d86c51d43b02de170b35200768473bd00cef31/apps/prairielearn/python/zygote.py#L272-L275

**Review by nwalters512 — APPROVED** [review]

File: apps/prairielearn/python/zygote.py (hunk 11 above)
Comment (reteps) [reply apps/prairielearn/python/zygote.py]: Removed in 2a955b47aaba80ca142730b9a9cdb229eddf92a8.

File: apps/prairielearn/python/README.md (hunk 12 above)
Comment (reteps) [reply apps/prairielearn/python/README.md]: Changed the text to reflect this distinction in 2a955b47aaba80ca142730b9a9cdb229eddf92a8.

File: apps/prairielearn/python/zygote.py [hunk 13]
```diff
@@ -66,8 +66,6 @@
 
 logging.getLogger("matplotlib.font_manager").disabled = True
 
-# Pre-load commonly used modules
```
Comment (nwalters512) [inline apps/prairielearn/python/zygote.py:69]: @reteps Please add this comment back. It refers to the block of unused imports that follow. This preloading is crucial for performance and deserves its own comment.
